tst/hooks/sdk/useMediaStreamMetrics.test.tsx

```diff
@@ -0,0 +1,191 @@
+    super.removeObserver(observer);
+  }
+}
+
+function createMeetingSessionConfiguration(): MeetingSessionConfiguration {
+  const configuration = new MeetingSessionConfiguration();
+  configuration.meetingId = 'foo-meeting';
+  configuration.urls = new MeetingSessionURLs();
+  configuration.urls.audioHostURL = 'https://audiohost.test.example.com';
+  configuration.urls.turnControlURL = 'https://turncontrol.test.example.com';
+  configuration.urls.signalingURL = 'https://signaling.test.example.com';
+  configuration.credentials = new MeetingSessionCredentials();
+  configuration.credentials.attendeeId = 'foo-attendee';
+  configuration.credentials.joinToken = 'foo-join-token';
+  configuration.attendeePresenceTimeoutMs = 5000;
+  return configuration;
+}
+
+function createClientMetricReport(): DefaultClientMetricReport {
+  class FakeClientMetricReport extends DefaultClientMetricReport {
+    constructor() {
+      super(new NoOpDebugLogger());
+    }
+
+    getObservableVideoMetrics(): { [p: string]: { [p: string]: {} } } {
+      return {
+        1: {
+          1: {
+            videoDownstreamBitrate: 8000,
+            videoDownstreamPacketLossPercent: 50,
+            videoDownstreamPacketsReceived: 1000,
+          },
+        },
+      };
+    }
+
+    getObservableMetrics(): { [p: string]: number } {
+      return {
+        availableOutgoingBitrate: 1000,
+        availableIncomingBitrate: 1000,
+      };
+    }
+  }
+
+  return new FakeClientMetricReport();
+}
+
+const initialMediaStreamMetrics = {
+  audioPacketsSentFractionLossPercent: null,
+  audioPacketsReceivedFractionLossPercent: null,
+  availableOutgoingBandwidth: null,
+  availableIncomingBandwidth: null,
+  videoStreamMetrics: {},
+};
+
+const expectedMediaStreamMetrics = {
+  audioPacketsSentFractionLossPercent: NaN,
+  audioPacketsReceivedFractionLossPercent: NaN,
+  availableOutgoingBandwidth: 1,
+  availableIncomingBandwidth: 1,
+  videoStreamMetrics: {
+    '1': {
+      '1': {
+        videoDownstreamBitrate: 8000,
+        videoDownstreamPacketLossPercent: 50,
+        videoDownstreamPacketsReceived: 1000,
+      },
+    },
+  },
+};
+
+function createAudioVideoFacade(): DefaultAudioVideoFacade {
+  class TestBackoff implements Backoff {
+    reset(): void {}
+
+    nextBackoffAmountMs(): number {
+      return 1;
+    }
+  }
+
+  const logger = new NoOpDebugLogger();
+  const deviceController = new DefaultAudioVideoController(
+    createMeetingSessionConfiguration(),
+    logger,
+    new DefaultWebSocketAdapter(logger),
+    new NoOpMediaStreamBroker(),
+    new DefaultReconnectController(120 * 1000, new TestBackoff())
+  );
+
+  // @ts-ignore
+  return new DefaultAudioVideoFacade(
+    deviceController,
+    // @ts-ignore
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    undefined
+  );
+}
+
+describe('useMediaStreamMetrics', () => {
+  it('mount/unmount', () => {
+    const audioVideoFacade = new NoOPAudioVideoFacade();
```

Comment:
If you already have a createAVFacade function then why do we need this anymore?
All in all I still think its abit complicated.
Can we just have a testAVFacade that extends `DefaultAVFacade` and just pass in `NoOPAVController`?  You can maintain your own observer queue via your overriden add/remove observer and add a function in testAVFacade to manually trigger `metricsDidReceive`.

Reply:
Make sense, will update it.

Reply:
I think the logic of getting metrics via `metricDidreceive` is already tested in JS SDK. Why we need to test it in React SDK?

Reply:
Updated!